Tests/SwiftFusionTests/Inference/GenericFactorTests.swift

```diff
@@ -19,57 +19,87 @@ import XCTest
 import PenguinStructures
 @testable import SwiftFusion
 
-/// A factor that selects a pose prior using an integer input index.
-fileprivate struct DemoFactor<Pose: LieGroup>: GenericFactor {
-  typealias Variables = Tuple2<Int, Pose>
+/// A factor that specifies a prior on a pose.
+fileprivate struct PriorFactor<Pose: LieGroup>: GenericFactor {
+  typealias Variables = Tuple1<Pose>
 
   let edges: Variables.Indices
 
-  let priors: [Pose]
+  let prior: Pose
 
   func error(_ variables: Variables) -> Double {
-    let (priorIndex, position) = (variables.head, variables.tail.head)
-    return priors[priorIndex].localCoordinate(position).squaredNorm
+    let actualPosition = variables.head
+    return prior.localCoordinate(actualPosition).squaredNorm
+  }
+}
+
+/// A factor that switches between different linear motions based on an integer label.
+fileprivate struct SwitchingLinearMotionFactor<Pose: LieGroup>: GenericFactor {
+  typealias Variables = Tuple3<Int, Pose, Pose>
+
+  let edges: Variables.Indices
+
+  let motions: [Pose]
+
+  func error(_ variables: Variables) -> Double {
+    let (motionLabel, start, end) = (variables.head, variables.tail.head, variables.tail.tail.head)
+    let actualMotion = between(start, end)
+    return motions[motionLabel].localCoordinate(actualMotion).squaredNorm
   }
 }
 
 class GenericFactorTests: XCTestCase {
   /// Tests errors in a simple factor graph.
   func testErrors() {
-    // Set up 4 variables:
-    // - `intVar1` is an `Int` with value `0`.
-    // - `intVar2` is an `Int` with value `1`.
-    // - `poseVar1` is a `Pose2` with value `Pose2(1, 0, 0)`.
-    // - `poseVar2` is a `Pose2` with value `Pose2(0, 1, 1)`.
+    // A switching linear motion factor graph.
+    //
+    // There are 3 pose variables and 2 motion labels in the graph. The initial guesses are:
+    // - motion labels: 0, 1
+    // - poses: Pose2(0, 0, 0), Pose2(1, 1, 0), Pose2(1, 1, 0.9)
+    //
+    // The factors are:
+    // - a prior factor giving a prior of `Pose2(0, 0, 0)` on the first position.
+    // - two motion factors with possible motions [Pose2(1, 1, 0), Pose2(0, 0, 1)]
+
+    // Set up the initial guess.
 
     let intVariables = ArrayStorage<Int>.create(minimumCapacity: 2)
-    let intVar1 = TypedID<Int, Int>(intVariables.append(0)!)
-    let intVar2 = TypedID<Int, Int>(intVariables.append(1)!)
+    let motionLabel1ID = TypedID<Int, Int>(intVariables.append(0)!)
```

Comment:
TypedID is so cool! But, wow, this syntax is a bit heavy

Reply:
This will be improved by one of the pieces that we haven't implemented yet -- there will be some container for variable assignments and adding a new variable will look something like this:

```
let motionLabel1ID = variableAssignments.store(0)
```